File: src/main/java/com/bazbatlabs/smashballs/models/SoundMap.java
package com.bazbatlabs.smashballs.models;

import java.util.*;
import java.io.*;

import android.content.res.Resources;
import android.media.AudioManager;
import android.media.SoundPool;

public final class SoundMap {

    private static final int MAX_STREAMS = 20;

    private final Map<Sound, Integer> map;
    private final SoundPool soundPool;

    public SoundMap(Resources resources) {
        this.soundPool = new SoundPool(MAX_STREAMS, AudioManager.STREAM_MUSIC, 0);
        this.map = new HashMap<Sound, Integer>();

        for (Sound sound : Sound.values()) {

            int id = -1;
            try {
                id = soundPool.load(resources.getAssets().openFd("sounds/" + sound.filename()), 1);

            } catch (IOException e) {
                throw new RuntimeException("Failed to load asset: " + sound.filename(), e);
            }

            map.put(sound, id);
        }
    }


    public static enum Sound {
        BLIP("blip.ogg"),
        BLIP_LOW("blip_low.ogg");

        private final String filename;

        Sound(String filename) {
            this.filename = filename;
        }

        public String filename() { return filename; }
    }

    public void play(Sound sound) {
        soundPool.play(map.get(sound), 1, 1, 0, 0, 1);
    }
}

File: src/main/scala/models/SoundMap.scala
package com.bazbatlabs.bricksmash.models

import java.io.IOException

import android.content.res.Resources
import android.media.AudioManager
import android.media.SoundPool

class SoundMap(resources: Resources) {
  val MaxStreams = 20

  val soundPool = new SoundPool(MaxStreams, AudioManager.STREAM_MUSIC, 0)

  val map = Map[Sound, Int](
    Blip -> loadSound(Blip.filename),
    BlipLow -> loadSound(BlipLow.filename)
  )

  def loadSound(filename: String): Int =
    try {
      soundPool.load(resources.getAssets.openFd("sounds/" + filename), 1)
    } catch {
      case ex: IOException => throw new RuntimeException("Failed to load asset: " + filename, ex)
    }

  def play(sound: Sound) {
    soundPool.play(map(sound), 1, 1, 0, 0, 1)
  }
}

sealed abstract class Sound(val filename: String)
case object Blip extends Sound("blip.ogg")
case object BlipLow extends Sound("blip_low.ogg")
